### ISSUE TITLE
Complex JSON Schema Validation Fail

### ISSUE THREAD

**Issue by Gourav2912** [Apicurio/apicurio-registry#2552]
Hello,

I have observed JSONSchema validation for complex object when object is inside an Array is not working as expected.
Validation of objects inside arrays are ignored by validation library.

I used attached JSON schema and register them in APICurio.
On validating below data, validation is successful which is supposed to fail as zipCode is string and expected to be integer.
```json
{
    "firstName":"568dfa8d-2cd9-473e-acb5-9be84382a74b",
    "lastName":"6635e20a-7be7-4b96-a5bc-c13fa57a4a04",
    "age":20,
    "city":[{"name":"573f459c-0aa4-4338-9821-60a04260cc7e","zipCode":"12345"}]
}
```

[citizen.zip](https://github.com/Apicurio/apicurio-registry/files/8844798/citizen.zip)

I sent the below data

**Comment by Gourav2912** [comment]
Hello, Is this issue discussed? Any update for this please?

**Comment by EricWittmann** [comment]
Can you explain the use-case further?  Are you using Apicurio SerDes with JSON Schema validation enabled in a Kafka application?  How are you validating the message/data using the schema registered in registry?

**Comment by Gourav2912** [comment]
Hello Eric...This is our scenario validating the message/data using the schema registered in registry.

**Comment by jsenko** [comment]
hi @Gourav2912 Could you please provide more details so we know exactly how to reproduce the problem? I've seen the data model you have attached, but can you provide the JSON schema you are using as well? What actions are you performing when the issue happens?

**Comment by Gourav2912** [comment]
Hello @jsenko . PFA small example project i created for issue. Project will register citizen and city schema in apicurio registry.
city is array object in citizen. In the example for data to be validated with register schema, city object is incorrect. Still on validating the data with apicurio SDK, validation is success. Hope this help to reproduce the issue
[apicurio-issue.zip](https://github.com/Apicurio/apicurio-registry/files/9146861/apicurio-issue.zip)
.

**Comment by jsenko** [comment]
Hello, thank you for the reproducer project, it was very helpful for diagnosing the issue. The reason why the validation failed was that we do not yet fully support references in serdes, so only the parent schema was used for validation. This is a new feature that we're still working on, but it should be ready in the near future. We can use you example for testing as well. CC @carlesarnal

**Comment by Gourav2912** [comment]
@jsenko @carlesarnal Request you to please check this issue , this is very common scenario and we want to have this ASAP to use APICurio in PROD

**Comment by carlesarnal** [comment]
If I'm getting this correctly, this is already solved by https://github.com/Apicurio/apicurio-registry/pull/2810. @Gourav2912 can you please check and eventually close?

**Comment by EricWittmann** [comment]
I was just going to post this same thing.  I suspect this is fixed. :)

**Comment by Gourav2912** [comment]
@EricWittmann @carlesarnal i guess jars not not yet release maven repo to test. If u can release 2.3.1 version with changes to test

**Comment by EricWittmann** [comment]
Hi @Gourav2912 - release of 2.3.1.Final is happening right now.  You should be able to give this a try soon (maybe in two hours or so unless the release workflow fails).

**Comment by Gourav2912** [comment]
@EricWittmann @carlesarnal Issue is solved partially yet, if there are more then 1 optional array object library is still expecting optional object to be present.
I have created sample project to re-create the issue. Please check example in class Validation_Standard. Use Create class to register schema in apicurio.
[apicurio-issue.zip](https://github.com/Apicurio/apicurio-registry/files/9643168/apicurio-issue.zip)

**Comment by Gourav2912** [comment]
@carlesarnal @EricWittmann By any chance did u get time to see my above request with project attached.

**Comment by EricWittmann** [comment]
Thanks for the reproducer, @Gourav2912 - we haven't had time to evaluate this yet.  Please hold.  :)

**Comment by carlesarnal** [comment]
I have a better sense of what is happening now. Actually, the validation library is not trying to validate the schema of the objects inside the array. When you get the data serialized from a java object, the array object is there, empty, and the validation library expects it to be there. In your reproducer, in the data being validated, the array is not there, so the validation library fails to check that. It's an easy fix, but, just so you're aware, you have that workaround already available that would make the library to work.

In any case, here's the PR that fixes the issue #2854.

Thanks

**Comment by carlesarnal** [comment]
Closing as it was confirmed in https://github.com/Apicurio/apicurio-registry/pull/2854 that the fix is working as expected.
